### ISSUE TITLE
How to pause with CountdownTimerController?

### ISSUE THREAD

**Issue by jrmallorca** [wuweijian1997/FlutterCountdownTimer#39]
Hello,
I've noticed that there are 2 controllers and from #14 you use CountdownController to pause. I can't see a method to pause using (what I'm speculating is the new one) CountdownTimerController.

**Comment by wuweijian1997** [comment]
You can use `controller.disposeTimer()` to pause it. But since `endTime` is a timestamp, it will skip the paused time when it starts again. See the `example` code for details.
```
  if (controller.isRunning) {
            controller.disposeTimer();
          } else {
            controller.start();
          }
```

**Comment by jrmallorca** [comment]
Thanks for the idea. I got the below code in the end. Unfortunately, playing again will work fine but visually skips a second. I don't know if you have a fix for that or a cleaner version of the below. I don't know if you'd like to close this issue now, thanks!
```
if (_controller.isRunning) {
  // Get the milliseconds
  int? milliseconds = _controller.currentRemainingTime!.milliseconds !=
          null
      ? _controller.currentRemainingTime!.milliseconds!.value.toInt()
      : 0;

  // Get the remaining time
  _remainingTime = Duration(
    days: _controller.currentRemainingTime!.days ?? 0,
    hours: _controller.currentRemainingTime!.hours ?? 0,
    minutes: _controller.currentRemainingTime!.min ?? 0,
    seconds: _controller.currentRemainingTime!.sec ?? 0,
    milliseconds: milliseconds,
  );

  _controller.disposeTimer();
} else {
  _timeToFinish =
      DateTime.now().add(_remainingTime).millisecondsSinceEpoch;

  _controller.endTime = _timeToFinish;
  _controller.start();
}
```

**Comment by wuweijian1997** [comment]
Is this what you need?There is also a plugin `Countdown` in the repository, which runs according to the given timing constants. You can see if this is the effect you are looking for.

```
import 'package:flutter/material.dart';
import 'package:flutter_countdown_timer/index.dart';

class CountdownPage extends StatefulWidget {
  @override
  _CountdownPageState createState() => _CountdownPageState();
}

class _CountdownPageState extends State<CountdownPage> {
  CountdownController countdownController =
      CountdownController(duration: Duration(hours: 1));
  bool isRunning = false;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Countdown(
            countdownController: countdownController,
            builder: (_, Duration time) {
              var currentRemainingTime = countdownController.currentRemainingTime;
              return Text(
                  'days: [ ${currentRemainingTime.days} ], hours: [ ${currentRemainingTime.hours} ], min: [ ${currentRemainingTime.min} ], sec: [ ${currentRemainingTime.sec} ]');
            }),
      ),
      floatingActionButton: FloatingActionButton(
        child:
            Icon(isRunning ? Icons.stop : Icons.play_arrow),
        onPressed: () {
          if (!countdownController.isRunning) {
            countdownController.start();
            setState(() {
              isRunning = true;
            });
          } else {
            countdownController.stop();
            setState(() {
              isRunning = false;
            });
          }
        },
      ),
    );
  }
}

```

**Comment by jrmallorca** [comment]
I'm trying this out right now and it seems to do the job great but only for pausing and playing. Calling dispose() gives me the first log and I don't know the proper way to dispose of it. I also can't find an appropriate way to call an onEnd() method, unlike CountdownTimerController. It causes an issue if I call it within the builder() method as I use setState in my onEnd() method.


<details>
  <summary>Code</summary>

dispose() exception:
```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following assertion was thrown while finalizing the widget tree:
A CountdownController was used after being disposed.
Once you have called dispose() on a CountdownController, it can no longer be used.

When the exception was thrown, this was the stack:
#0      ChangeNotifier._debugAssertNotDisposed.<anonymous closure> (package:flutter/src/foundation/change_notifier.dart:117:9)
#1      ChangeNotifier._debugAssertNotDisposed (package:flutter/src/foundation/change_notifier.dart:123:6)
#2      ChangeNotifier.dispose (package:flutter/src/foundation/change_notifier.dart:212:12)
#3      CountdownController.dispose (package:flutter_countdown_timer/countdown_controller.dart:90:11)
#4      _PreparationPageState.dispose (package:habitualise/pages/exercise/preparation_page.dart:181:17)
#5      StatefulElement.unmount (package:flutter/src/widgets/framework.dart:4721:11)
#6      _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2022:13)
#7      _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#8      SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:5961:14)
#9      _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#10     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#11     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#12     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#13     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#14     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:5961:14)
#15     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#16     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#17     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:5961:14)
#18     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#19     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#20     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#21     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#22     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#23     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:5961:14)
#24     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#25     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#26     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:5961:14)
#27     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#28     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#29     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#30     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#31     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#32     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#33     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#34     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#35     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#36     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#37     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#38     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:5961:14)
#39     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#40     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#41     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#42     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#43     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#44     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:5961:14)
#45     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#46     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#47     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#48     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#49     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#50     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#51     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#52     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#53     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#54     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#55     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#56     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#57     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#58     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#59     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#60     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#61     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#62     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#63     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#64     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#65     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:5961:14)
#66     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#67     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#68     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#69     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#70     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#71     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#72     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#73     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#74     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#75     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#76     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#77     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#78     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#79     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#80     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#81     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#82     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#83     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:5961:14)
#84     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#85     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#86     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#87     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#88     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#89     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#90     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#91     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2020:7)
#92     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4549:14)
#93     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2018:13)
#94     ListIterable.forEach (dart:_internal/iterable.dart:39:13)
#95     _InactiveElements._unmountAll (package:flutter/src/widgets/framework.dart:2031:25)
#96     BuildOwner.finalizeTree.<anonymous closure> (package:flutter/src/widgets/framework.dart:2785:27)
#97     BuildOwner.lockState (package:flutter/src/widgets/framework.dart:2585:15)
#98     BuildOwner.finalizeTree (package:flutter/src/widgets/framework.dart:2784:7)
#99     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:875:19)
#100    RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:319:5)
#101    SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1144:15)
#102    SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1082:9)
#103    SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:998:5)
#107    _invoke (dart:ui/hooks.dart:161:10)
#108    PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:253:5)
#109    _drawFrame (dart:ui/hooks.dart:120:31)
(elided 3 frames from dart:async)
════════════════════════════════════════════════════════════════════════════════════════════════════

```

setState() exception:
```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following assertion was thrown building Countdown(dirty, state: _CountdownState#266fd):
setState() or markNeedsBuild() called during build.
This PreparationPage widget cannot be marked as needing to build because the framework is already in
the process of building widgets.  A widget can be marked as needing to be built during the build
phase only if one of its ancestors is currently building. This exception is allowed because the
framework builds parent widgets before children, which means a dirty descendant will always be
built. Otherwise, the framework might not visit this widget during this build phase.
The widget on which setState() or markNeedsBuild() was called was:
  PreparationPage
The widget which was currently being built when the offending call was made was:
  Countdown

The relevant error-causing widget was:
  Countdown
  file:///home/jrmallorca/Projects/individualProject/application/habitualise/lib/pages/exercise/preparation_page.dart:282:12

When the exception was thrown, this was the stack:
#0      Element.markNeedsBuild.<anonymous closure> (package:flutter/src/widgets/framework.dart:4138:11)
#1      Element.markNeedsBuild (package:flutter/src/widgets/framework.dart:4153:6)
#2      State.setState (package:flutter/src/widgets/framework.dart:1287:15)
#3      _PreparationPageState._onFinishCountdown (package:habitualise/pages/exercise/preparation_page.dart:143:9)
#4      _PreparationPageState._buildCountdownTimer.<anonymous closure> (package:habitualise/pages/exercise/preparation_page.dart:285:36)
#5      _CountdownState.build (package:flutter_countdown_timer/countdown.dart:46:20)
#6      StatefulElement.build (package:flutter/src/widgets/framework.dart:4612:27)
#7      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4495:15)
#8      StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4667:11)
#9      Element.rebuild (package:flutter/src/widgets/framework.dart:4189:5)
#10     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2694:33)
#11     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:873:21)
#12     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:319:5)
#13     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1144:15)
#14     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1082:9)
#15     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:998:5)
#19     _invoke (dart:ui/hooks.dart:161:10)
#20     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:253:5)
#21     _drawFrame (dart:ui/hooks.dart:120:31)
(elided 3 frames from dart:async)

════════════════════════════════════════════════════════════════════════════════════════════════════

```
</details>

**Comment by wuweijian1997** [comment]
I will update a version to try to fix these issues.

**Comment by wuweijian1997** [comment]
I have released `4.1.0` version, you can see if it can solve your problem.

[example](https://github.com/wuweijian1997/FlutterCountdownTimer/blob/master/example/lib/countdown_page.dart)

**Comment by jrmallorca** [comment]
Thank you for implementing an `onEnd()` method! It's working now but I'm still confused as to how to `dispose()` properly as it keeps saying that ` Another exception was thrown: A CountdownController was used after being disposed. ` just like my above comment with the logs.

This happens when I call `Navigator.pop(context)` to exit from where `Countdown` is displayed. I've already overriden my `dispose()` such that it's like this:

```
@override
void dispose() {
  _controller.dispose();
  super.dispose();
}
```

Edit: I've changed it to:

```
void _handleFinish() {
  _controller.stop();
  Navigator.pop(context);
}
```

and removed the overriden `dispose()` method and it doesn't complain anymore about `CountdownController` being used. I don't know if that's an issue but I've fixed all of mine for now. Thank you for you help!
